### plot_snv_tool_summary.R ########################################################################
# Plot overlap of all availabe tools

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

        # build up the filename
        file.name <- paste(project.stem, file.core, sep = '_');
        file.name <- paste(file.name, extension, sep = '.');

        if (include.date) {
                file.name <- paste(Sys.Date(), file.name, sep = '_');
                }

        return(file.name);
        }

# function to write session profile to file
save.session.profile <- function(file.name) {

        # open the file
        sink(file = file.name, split = FALSE);

        # write memory usage to file
        cat('### MEMORY USAGE ###############################################################');
        print(proc.time());

        # write sessionInfo to file
        cat("\n### SESSION INFO ###############################################################");
        print(sessionInfo());

        # close the file
        sink();

        }

# function to trim sample IDs
simplify.ids <- function(x) {
	match <- TRUE;
	if (length(x) == 1) {
		match <- FALSE;
		new.ids <- x;
		}
	index <- 1;
	while (match) {
		if (length(unique(sapply(x,function(i) { unlist(strsplit(i,''))[index] } ))) == 1) {
			index <- index + 1;
			} else {
			new.ids <- sapply(x,function(i) { substring(i,index,nchar(i)) } );
			match <- FALSE;
			}
		}
	return(new.ids);
	}

### PREPARE SESSION ################################################################################
# import command line arguments
library(argparse);

parser <- ArgumentParser();

parser$add_argument('-p', '--project', type = 'character', help = 'PROJECT name');
parser$add_argument('-o', '--output', type = 'character', help = 'path to output directory');
parser$add_argument('-i', '--input', type = 'character',
	help = 'path to RData object generated by format_ensemble_mutations.R');
parser$add_argument('-z', '--report', type = 'character', help = 'path to report directory',
	default = NULL);

arguments <- parser$parse_args();

# import libraries
library(BoutrosLab.plotting.general);
library(UpSetR);

### READ DATA ######################################################################################
load(arguments$input);

if (!'combined.data' %in% ls()) {
	stop(paste0("STOP: Object 'combined.data' not found in ", arguments$input));
	}

# create (if necessary) and move to output directory
if (!dir.exists(arguments$output)) {
	dir.create(arguments$output);
	}

setwd(arguments$output);

### FORMAT DATA ####################################################################################
all.tools <- c('MuTect2','MuTect','Strelka','VarScan','SomaticSniper','VarDict','Pindel');
tool.list <- c();
for (tool in all.tools) {
	if (tool %in% colnames(combined.data)) { tool.list <- c(tool.list, tool); }
	}

# do some quick error checks
tool.count <- length(tool.list);

if (tool.count < 2) {
	stop("STOP: Data only available for a single tool - can't plot tool overlap summary.");
	}

# ensure sample IDs are a factor
all.samples <- unique(as.character(combined.data$Tumor_Sample_Barcode));
combined.data$Tumor_Sample_Barcode <- factor(combined.data$Tumor_Sample_Barcode,
	levels = sort(all.samples)
	);

# initiate plot data
plot.data <- list();

# get per-tool, per-sample mutation counts
plot.data$per_tool <- aggregate(
	combined.data[,tool.list],
	by = list(Tumor_Sample_Barcode = combined.data$Tumor_Sample_Barcode),
	FUN = function(i) { length(i[!is.na(i)]) }
	);

# find overlap
plot.data$overlap <- merge(
	aggregate(Chromosome ~ Tumor_Sample_Barcode + Count, combined.data, length),
	aggregate(Chromosome ~ Tumor_Sample_Barcode, combined.data, length),
	by = 'Tumor_Sample_Barcode'
	);
colnames(plot.data$overlap) <- c('ID','Overlap','Count','Total');
plot.data$overlap$Overlap <- factor(plot.data$overlap$Overlap, levels = seq(1:tool.count));
plot.data$overlap$Percent <- plot.data$overlap$Count / plot.data$overlap$Total * 100;

# save for future plotting
save(
	plot.data,
	file = generate.filename(arguments$project, 'mutation_overlap', 'RData')
	);

# now format data for plotting (UPSET plot)
tool.data <- combined.data[,tool.list];
tool.data[is.na(tool.data)] <- 0;
tool.data <- sign(tool.data);

### PLOT DATA ######################################################################################
total.combinations <- 0;
for (i in 1:tool.count) { total.combinations <- total.combinations + length(combn(tool.count, m = i)) }

# plot tool summary (overlap plot)
png(filename = generate.filename(arguments$project, 'SNV_tool_overlap','png'),
	res = 200, units = 'in', height = 5, width = 7);

upset(tool.data, sets = tool.list, mainbar.y.label = 'Number of Variants',
	sets.x.label = 'Number of Variants', show.numbers = FALSE,
	nintersects = if (total.combinations < 100) { NA } else { 50 }); #, set_size.scale_max = 85000);

dev.off();

# grab some parameters
axis.cex <- if (nrow(plot.data$per_tool) <= 30) { 1
	} else if (nrow(plot.data$per_tool) <= 50) { 0.75
	} else if (nrow(plot.data$per_tool) <= 80) { 0.5
	} else { 0 };

# create some plots to summarize overlap
plot.objects <- list();

# plot per-tool counts
for (tool in tool.list) {

	plot.objects[[tool]] <- create.barplot(
		get(tool) ~ Tumor_Sample_Barcode | tool,
		plot.data$per_tool,
		ylab.label = NULL,
		xlab.label = NULL,
		yaxis.cex = 0.8,
		xaxis.lab = rep('',nrow(plot.data$per_tool)),
		yaxis.tck = c(0.5,0),
		xaxis.tck = 0,
		yaxis.fontface = 'plain'
		);
	}

# create a legend (up to max tools used)
overlap.colour.scheme <- c(default.colours(6,'seq.greenblue'),'black');
overlap.legend <- legend.grob(
	legends = list(
		legend = list(
			colours = rev(overlap.colour.scheme[1:tool.count]),
			labels = rev(as.character(seq(1:tool.count))),
			title = 'Tool Count'
			)
		),
	title.just = 'left',
	size = 2
	);

# plot overlap
plot.objects[['overlap']] <- create.barplot(
	Percent ~ ID,
	plot.data$overlap,
	groups = plot.data$overlap$Overlap,
	stack = TRUE,
	col = overlap.colour.scheme[1:tool.count], 
	ylab.label = '% of Total',
	xlab.label = NULL,
	ylimits = c(0,100),
	yat = seq(0,100,20),
	ylab.cex = 1,
	yaxis.cex = 1,
	xaxis.lab = simplify.ids(levels(plot.data$overlap$ID)),
	xaxis.cex = axis.cex,
	yaxis.tck = c(0.5,0),
	xaxis.tck = if (axis.cex == 0) { 0 } else { c(0.5, 0) },
	xaxis.rot = 90,
	yaxis.fontface = 'plain',
	xaxis.fontface = 'plain',
	legend = list(
		right = list(fun = overlap.legend)
		)			
	);

# combine them!
create.multipanelplot(
	plot.objects = plot.objects,
	height = 11,
	width = 8,
	resolution = 200,
	filename = generate.filename(arguments$project, 'mutation_overlap','png'),
	plot.objects.heights = c(rep(1,tool.count),2.5),
	left.legend.padding = 0,
	right.legend.padding = 0,
	top.legend.padding = 0,
	bottom.legend.padding = 0,
	y.spacing = c(rep(-2.5,6),-1),
	ylab.label = paste(
		paste(rep(' ', 50), collapse = ''),
		'Somatic Variant Count (SNVs and INDELs)'
		),
	ylab.cex = 1,
	left.padding = 0,
	ylab.axis.padding = 0,
	bottom.padding = 0,
	top.padding = -1,
	right.padding = 0
	);

# run unlink, in case it exists from a previous run
if (!is.null(arguments$report)) {
	unlink(paste0(arguments$report, '/', 'SNV_tool_overlap.png'));
	file.symlink(
		paste0(arguments$output, '/',
			generate.filename(arguments$project, 'SNV_tool_overlap','png')),
		paste0(arguments$report, '/', 'SNV_tool_overlap.png')
		);

	unlink(paste0(arguments$report, '/', 'mutation_tool_overlap.png'));
	file.symlink(
		paste0(arguments$output, '/',
			generate.filename(arguments$project, 'mutation_overlap','png')),
		paste0(arguments$report, '/', 'mutation_tool_overlap.png')
		);
	}

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('SNVToolOverlap','SessionProfile','txt'));
