### plot_methylation_summary.R #####################################################################
# Plot methylation landscape.
# INPUT:
#	- gene-wise methylation values (output by collect_methyldackel_output.R)

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

# function to trim sample IDs
simplify.ids <- function(x) {
	match <- TRUE;
	if (length(x) == 1) {
		match <- FALSE;
		new.ids <- x;
		}
	index <- 1;
	while (match) {
		if (length(unique(sapply(x,function(i) { unlist(strsplit(i,''))[index] } ))) == 1) {
			index <- index + 1;
			} else {
			new.ids <- sapply(x,function(i) { substring(i,index,nchar(i)) } );
			match <- FALSE;
			}
		}
	return(new.ids);
	}

### PREPARE SESSION ################################################################################
# import command line arguments
library(argparse);

parser <- ArgumentParser();

parser$add_argument('-p', '--project', type = 'character', help = 'PROJECT name');
parser$add_argument('-o', '--output', type = 'character', help = 'path to output directory');
parser$add_argument('-m', '--methylation', type = 'character', 
	help = 'path to gene by sample matrix of methylation levels (generated by collect_methyldackel_output.R)');
parser$add_argument('-r', '--regulatory', type = 'character', 
	help = 'path to promoter by sample matrix of methylation levels (generated by collect_methyldackel_output.R)');
parser$add_argument('-z', '--report', type = 'character', help = 'path to report directory',
	default = NULL);

arguments <- parser$parse_args();

# import libraries
library(xtable);
library(BoutrosLab.plotting.general);

### READ DATA ######################################################################################
# get data
if (is.null(arguments$methylation)) {
	stop('ERROR: No input file provided, please provide path to MethylDackel results matrix');
	} else {
	gene.data <- read.delim(arguments$methylation, stringsAsFactors = FALSE);
	}

if (is.null(arguments$regulatory)) {
	stop('ERROR: No input file provided, please provide path to MethylDackel results matrix');
	} else {
	promoter.data <- read.delim(arguments$regulatory, stringsAsFactors = FALSE);
	}

# create (if necessary) and move to output directory
if (!dir.exists(arguments$output)) {
	dir.create(arguments$output);
	}

setwd(arguments$output);

### FORMAT DATA ####################################################################################
# collect list of all (ordered) samples
smp.idx <- if ('TXNAME' %in% colnames(gene.data)) {
	9:ncol(gene.data); } else { 7:ncol(gene.data); }

# do some initial filtering
gene.data <- gene.data[!is.na(gene.data$ENTREZID),];
promoter.data <- promoter.data[!is.na(promoter.data$ENTREZID),];

# split gene.data
anno.data <- gene.data[,-smp.idx];

plot.data <- gene.data[,smp.idx];
rownames(plot.data) <- gene.data$SYMBOL;
colnames(plot.data) <- gsub('\\.','-', colnames(plot.data));


# remove low-qc samples
na.counts <- apply(plot.data,2,function(i) { length(i[is.na(i)])/length(i) } );
remove.smps <- names(na.counts)[which(na.counts > 0.5)];

plot.data <- plot.data[,!colnames(plot.data) %in% remove.smps];

# do some minor filtering to remove genes with no variation across the cohort
na.counts <- apply(plot.data,1,function(i) { length(i[is.na(i)]) } );
remove.genes <- names(na.counts)[which(na.counts > ncol(plot.data)*0.6)];

plot.data <- plot.data[!rownames(plot.data) %in% remove.genes,];

# calculate variance across the samples
gene.variances <- apply(plot.data,1,var,na.rm = TRUE);
if (nrow(plot.data) > 20000) {
	plot.data <- plot.data[which(gene.variances > 1),];
	}


# subset promoter data to MMR genes
mmr.genes <- c('MLH1','MLH3','MSH2','MSH3','MSH6','PMS2');
plot.data.mmr <- promoter.data[which(promoter.data$SYMBOL %in% mmr.genes),];
rownames(plot.data.mmr) <- plot.data.mmr$SYMBOL;
plot.data.mmr <- plot.data.mmr[,smp.idx];
colnames(plot.data.mmr) <- gsub('\\.','-', colnames(plot.data.mmr));

plot.data.mmr <- plot.data.mmr[,!colnames(plot.data.mmr) %in% remove.smps];

### PLOT DATA ######################################################################################
# grab some parameters
axis.cex <- if (ncol(plot.data) <= 30) { 1
	} else if (ncol(plot.data) <= 50) { 0.75
	} else if (ncol(plot.data) <= 80) { 0.5
	} else if (ncol(plot.data) <= 100) { 0
	} else { 0 };

# create heatmap
create.heatmap(
	t(plot.data),
	cluster.dimensions = 'rows',
	xaxis.lab = simplify.ids(colnames(plot.data)),
	yaxis.lab = rep('', nrow(plot.data)),
	xaxis.cex = axis.cex,
	yaxis.tck = 0,
	xaxis.tck = 0.2,
	xaxis.fontface = 'plain',
	axes.lwd = 1,
	print.colour.key = TRUE,
	fill.colour = 'grey70',
	at = seq(0,1,0.01),
	colour.scheme = c('magenta3','black','darkolivegreen1'),
	colourkey.cex = 1,
	height = 8,
	width = 10,
	resolution = 200,
	filename = generate.filename(arguments$project, 'methylation_landscape','png')
	);

# subset to most variable genes
gene.order <- names(rev(sort(gene.variances)));
top.genes <- gene.order[1:25];

plot.data.top <- plot.data[top.genes,];

create.heatmap(
	t(plot.data.top),
	cluster.dimensions = 'both',
	xaxis.lab = simplify.ids(colnames(plot.data.top)), 
	yaxis.lab = NA,
	xaxis.cex = axis.cex,
	yaxis.cex = 0.9,
	xaxis.tck = 0.2,
	yaxis.tck = 0.2,
	xaxis.fontface = 'plain',
	yaxis.fontface = 'italic',
	axes.lwd = 1,
	print.colour.key = TRUE,
	fill.colour = 'grey70',
	at = seq(0,1,0.01),
	colour.scheme = c('magenta3','black','darkolivegreen1'),
	colourkey.cex = 1,
	height = 6,
	width = 10,
	resolution = 200,
	filename = generate.filename(arguments$project, 'methylation_landscape__topVariance','png')
	);


summary.metrics <- merge(
	anno.data[anno.data$SYMBOL %in% top.genes,c('SYMBOL','ENTREZID')],
	data.frame(
		Mean = apply(plot.data.top,1,mean, na.rm = TRUE),
		Median = apply(plot.data.top,1,median, na.rm = TRUE),
		Lower_25 = apply(plot.data.top,1,quantile, p = 0.25, na.rm = TRUE),
		Upper_75 = apply(plot.data.top,1,quantile, p = 0.75, na.rm = TRUE),
		SD = apply(plot.data.top,1,sd, na.rm = TRUE)
		),
	by.x = 'SYMBOL',
	by.y = 'row.names'
	);

summary.metrics$SYMBOL <- factor(summary.metrics$SYMBOL, levels = top.genes);
summary.metrics <- summary.metrics[order(summary.metrics$SYMBOL),];


# subset to key MMR gene promoters
create.heatmap(
	t(plot.data.mmr),
	cluster.dimensions = 'none',
	xaxis.lab = simplify.ids(colnames(plot.data.mmr)), 
	yaxis.lab = NA,
	xaxis.cex = axis.cex,
	yaxis.cex = axis.cex,
	xaxis.tck = 0.2,
	yaxis.tck = 0.2,
	xaxis.fontface = 'plain',
	yaxis.fontface = 'italic',
	axes.lwd = 1,
	print.colour.key = TRUE,
	fill.colour = 'grey70',
	at = seq(0,1,0.01),
	colour.scheme = c('magenta3','black','darkolivegreen1'),
	colourkey.cex = 1,
	height = 4,
	width = 10,
	resolution = 200,
	filename = generate.filename(arguments$project, 'methylation_landscape__MMR_promoters','png')
	);
	

### LATEX ##########################################################################################
# if making output for a report
if (!is.null(arguments$report)) {

	# initiate report object
	tex.file <- paste0(arguments$report, '/methylation_summary.tex');

	# write some captions
	full.caption <- "Gene-wise methylation profile; values are percent of reads with methyl-C.";
	trimmed.caption <- "Methylation profile (fraction of reads with methyl-C) for the top most variable genes.";
	mmr.caption <- "Methylation profile (fraction of reads with methyl-C) for mismatch repair gene promoters.";

	# write for latex
	write("\\section{Methylation Levels}", file = tex.file);

	# full expression profile
	write("\\begin{figure}[h!]", file = tex.file, append = TRUE);
	write("\\begin{center}", file = tex.file, append = TRUE);
	write(paste0(
		"\\includegraphics[width=0.9\\textwidth]{",
		getwd(), '/',
		generate.filename(arguments$project, 'methylation_landscape__topVariance','png'), '}'
		), file = tex.file, append = TRUE);
	write("\\end{center}", file = tex.file, append = TRUE);
	write(paste0(
		"\\caption{", trimmed.caption, "}"
		), file = tex.file, append = TRUE);
	write("\\end{figure}\n\\pagebreak\n", file = tex.file, append = TRUE);

	print(
		xtable(
			summary.metrics[,c('SYMBOL','Lower_25','Median','Upper_75')],
			caption = 'Top 25 most variably methylated genes across the cohort, ordered by decreasing variance. Table shows gene symbol, lower 25th percentile, median and upper 25th percentile of expression values.'
			),
		file = tex.file,
		include.rownames = FALSE,
		table.placement = 'h!',
		append = TRUE
		);

	# MMR gene promoters
	write("\\pagebreak\n\\begin{figure}[h!]", file = tex.file, append = TRUE);
	write("\\begin{center}", file = tex.file, append = TRUE);
	write(paste0(
		"\\includegraphics[width=0.9\\textwidth]{",
		getwd(), '/',
		generate.filename(arguments$project, 'methylation_landscape__MMR_promoters','png'), '}'
		), file = tex.file, append = TRUE);
	write("\\end{center}", file = tex.file, append = TRUE);
	write(paste0(
		"\\caption{", mmr.caption, "}"
		), file = tex.file, append = TRUE);
	write("\\end{figure}", file = tex.file, append = TRUE);

	}

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('MethylationSummary','SessionProfile','txt'));
