### collect_ascat_output.R #########################################################################
# Finds and combines output generated by ASCAT (CN segments, cellularity, ploidy)

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

### PREPARE SESSION ################################################################################
# import libraries
library(GenomicRanges);
library(org.Hs.eg.db);
library(argparse);

# import command line arguments
parser <- ArgumentParser();

parser$add_argument('-d', '--directory', type = 'character', help = 'path to data directory');
parser$add_argument('-p', '--project', type = 'character', help = 'project name');
parser$add_argument('-r', '--ref_type', type = 'character', help = 'reference type', default = 'hg38');

arguments <- parser$parse_args();

# what's the date?
date <- Sys.Date();

setwd(arguments$directory);

### FORMAT ANNOTATION ##############################################################################
if (arguments$ref_type %in% c('hg38','GRCh38')) {
	library(TxDb.Hsapiens.UCSC.hg38.knownGene);
	txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene;
	} else if (arguments$ref_type %in% c('hg19','GRCh37')) {
	library(TxDb.Hsapiens.UCSC.hg19.knownGene);
	txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene;
	}
 
# organize all transcripts by GeneID
txdb_table <- transcriptsBy(txdb, by = 'gene');

# extract GeneIDs
tx_ids <- names(txdb_table);

# extract 
transcript_table <- select(
	TxDb.Hsapiens.UCSC.hg38.knownGene,
	keys = tx_ids,
	columns = c('GENEID','TXNAME','TXCHROM','TXSTART','TXEND'),
	keytype = 'GENEID'
	);
colnames(transcript_table)[1] <- 'ENTREZID';

# pull in extra annotations
transcript_table$SYMBOL <- mapIds(
	org.Hs.eg.db,
	keys = transcript_table$TXNAME,
	keytype = 'UCSCKG',
	column = 'SYMBOL'
	);

transcript_table$ENSEMBL <- mapIds(
	org.Hs.eg.db,
	keys = transcript_table$TXNAME,
	keytype = 'UCSCKG',
	column = 'ENSEMBL'
	);

transcript_table$MAP <- mapIds(
	org.Hs.eg.db,
	keys = transcript_table$TXNAME,
	keytype = 'UCSCKG',
	column = 'MAP'
	);

transcript_table$GENETYPE <- mapIds(
	org.Hs.eg.db,
	keys = transcript_table$TXNAME,
	keytype = 'UCSCKG',
	column = 'GENETYPE'
	);

# remove unmatching cases
transcript_table$MAPCHROM <- sapply(
	transcript_table$MAP,
	function(i) { paste0('chr', unlist(strsplit(i, 'p|q'))[1]) }
	);

transcript_table <- transcript_table[which(transcript_table$TXCHROM == transcript_table$MAPCHROM),];

# squish it to 1 entry per gene
gene.annotation <- merge(
	aggregate(TXSTART ~ ENTREZID + SYMBOL + ENSEMBL + GENETYPE + TXCHROM, transcript_table, min),
	aggregate(TXEND ~ ENTREZID + SYMBOL + ENSEMBL + GENETYPE + TXCHROM, transcript_table, max),
	);
colnames(gene.annotation)[5:7] <- c('Chromosome','Start','End');
gene.annotation$Chromosome <- factor(gene.annotation$Chromosome, levels = paste0('chr',c(1:22,'X','Y')));
gene.annotation <- gene.annotation[order(gene.annotation$Chromosome, gene.annotation$Start),];

gene.gr <- GRanges(gene.annotation);

### MAIN ###########################################################################################
# find results files
metric.files <- list.files(pattern = 'ASCAT_purity_ploidy.txt', recursive = TRUE);
cn.files <- list.files(pattern = 'segments.txt', recursive = TRUE);
gistic.files <- list.files(pattern = 'Total_CN.seg', recursive = TRUE);

# read them in
metric.list <- list();
cn.list <- list();
ratio.list <- list();

# read in metrics
for (file in metric.files) {
	# extract sample ID
	smp <- unlist(strsplit(basename(file), '_ASCAT'))[1];
	# store data in list
	metric.list[[smp]] <- read.delim(file)[,1:4];
	}

metric.data <- do.call(rbind, metric.list);
rownames(metric.data) <- 1:nrow(metric.data);
metric.data$PGA <- NA;

# read in CN calls
for (file in cn.files) {
	# extract sample ID
	smp <- unlist(strsplit(basename(file), '_absolute'))[1];
	# store data in list
	tmp <- read.delim(file);
	tmp$ID <- smp;
	cn.list[[smp]] <- tmp;
	rm(tmp);
	}

# read in logR data
for (file in gistic.files) {
	# extract sample ID
	smp <- unlist(strsplit(basename(file), '_Total'))[1];
	# store data in list
	ratio.list[[smp]] <- read.delim(file);
	}

cn.data <- do.call(rbind, cn.list);
ratio.data <- do.call(rbind, ratio.list);

# save combined/formatted data to file
write.table(
	ratio.data,
	file = generate.filename(arguments$project, 'ascat_segments_for_cbioportal', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	quote = FALSE,
	sep = '\t'
	);

# add chr prefix if necessary
if (!grepl('chr',ratio.data$chrom[1])) {
	ratio.data$chrom <- paste0('chr',ratio.data$chrom);
	}

# save combined/formatted data to file
write.table(
	ratio.data,
	file = generate.filename(arguments$project, 'ascat_segments_for_gistic', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	quote = FALSE,
	sep = '\t'
	);

# define a results table
gene.data.cn <- gene.annotation;
gene.data.cn[,names(cn.list)] <- 0;
gene.data.ratio <- gene.data.cn;

for (smp in names(cn.list)) {

	# convert ratios to GRanges
	tmp.abs <- ratio.list[[smp]][,c('chrom','loc.start','loc.end','seg.mean')];
	colnames(tmp.abs) <- c('chrom','start','end','RATIO');

	# convert CN calls to GRanges
	tmp.cn <- cn.list[[smp]][,c('chromosome','start.pos','end.pos','CNt')];
	colnames(tmp.cn) <- c('chrom','start','end','CN');
	tmp.cn$CN <- tmp.cn$CN - 2;

	# calculate PGA
	for.pga <- tmp.cn[which(tmp.cn$CN != 0),];
	pga <- sum(for.pga$end - for.pga$start)/(3*10**9)*100;
	metric.data[which(metric.data$Sample == smp),]$PGA <- pga;

	# combine ratio and CN calls
	abs.gr <- GRanges(tmp.abs);
	cn.gr <- GRanges(tmp.cn);

	tmp <- data.frame(mergeByOverlaps(abs.gr, cn.gr));
	tmp <- tmp[,c('abs.gr.seqnames','abs.gr.start','abs.gr.end','RATIO','CN')];
	colnames(tmp)[1:3] <- c('chrom','start','end');
	rownames(tmp) <- paste0('seg', 1:nrow(tmp));
	if (!grepl('chr',tmp$chrom[1])) {
		tmp$chrom <- paste0('chr',tmp$chrom);
		}
	seg.gr <- GRanges(tmp);

	# find overlap with genes (any amount of overlap)
	overlapGenes <- as.data.frame(findOverlaps(seg.gr, gene.gr, minoverlap = 20));
	regionsWithHits <- as.data.frame(seg.gr[unique(overlapGenes$queryHits)]);

	for (i in rownames(regionsWithHits)) {
		genes <- gene.gr[overlapGenes[which(overlapGenes$queryHits == sub('seg','',i)),]$subjectHits,]$SYMBOL;
		gene.data.cn[which(gene.data.cn$SYMBOL %in% genes),smp] <- tmp[i,]$CN;
		gene.data.ratio[which(gene.data.ratio$SYMBOL %in% genes),smp] <- tmp[i,]$RATIO;
		}
	}

# save data to file
write.table(
	metric.data,
	file = generate.filename(arguments$project, 'ascat_purity_ploidy','tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

write.table(
	gene.data.cn,
	file = generate.filename(arguments$project, 'ascat_cna_gene_matrix', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

write.table(
	gene.data.ratio,
	file = generate.filename(arguments$project, 'ascat_ratio_gene_matrix', 'tsv'),
	row.names = FALSE,
	col.names = TRUE,
	sep = '\t'
	);

### SAVE SESSION INFO ##############################################################################
save.session.profile(generate.filename('CollectCNAs','SessionProfile','txt'));
